### PR TITLE
feat(fc): Initial wiring of the flow control layer

### LINKED ISSUE #674: Introduce EPP-Level Queuing/Fairness for InferenceModels
### Current State
Currently, the Gateway API Inference Extension handles requests by directly invoking backend selection logic for each incoming request at a point in time as they are received. This results in a First-Come-First-Serve (FCFS) dispatch order based on request arrival, with backend choice dependent on instantaneous pool state (LoRA affinity, KV cache, backend queue lengths, etc.).

### Limitations
This direct dispatch model lacks mechanisms to:

1. **Guarantee Criticality-Based Service Differentiation:** Cannot ensure `Critical` requests are *always* dispatched before `Default` or `Sheddable` ones when InferenceModels are competing over limited pool resources.
2. **Enforce Inter-Model (Dispatch/Fairness) Policies:** Lacks a central point to apply dispatch policies (e.g., FCFS, Round Robin or other fairness definitions) *between* different `InferenceModels` of the same criticality.
3. **Optimally Handle Contention/Saturation:** Forces an immediate backend choice or request drop decision, potentially leading to suboptimal load distribution or unnecessary request failures when backends are only temporarily busy.

### Feature Request: Centrally Queue Requests in the EPP before the Backend Selection Decision

Introduce a **Queuing/Fairness Layer** before the backend selection step. This involves:

- Queuing requests per `InferenceModel`, grouped by `criticality`.
- Implementing an inter-criticality dispatch mechanism ensuring **strict priority** (Critical > Default > Sheddable).
- Allowing **pluggable inter-model dispatch policies** (e.g., FCFS, Round Robin, or other fairness definitions) *within* a priority band to manage inter-model fairness.
- Basic queue management (TTL, limits)

### Why Add Another Layer of Queuing?

- **Enforces Criticality:** Guarantees priority *before* resources are committed via backend selection.
- **Enables Fairness Policies:** Provides the necessary state and control point for inter-model dispatch logic.
- **Improved Contention Management:**
    - Decouples *which* request is dispatched next from *where* it goes.
    - Allows requests to wait for better backend states instead of immediate suboptimal dispatch or dropping.
    - Potentially improves load distribution across the pool by considering the *global* demand of pending requests when dispatching, reducing the chance of overloading specific backends.
    - Supports more intelligent back pressure signals for upstream components.
    - Shifts Head-of-Line (HoL) Blocking: While HoL blocking can occur in model server queues today, introducing a EPP-level queue shifts this potential blocking point "left". The benefit is that when a request at the head of the EPP queue is dispatched, the system *retains the flexibility to choose the best available backend* from the entire pool at that moment, rather than having committed the request prematurely to a specific, potentially suboptimal, model server queue where it might block others.
- **Potential Performance Gains:** By making more informed dispatch decisions (enabled by waiting and the global view) and improving load distribution (enabled by better backend selection flexibility for HoL requests), this approach *may improve tail latency and overall throughput*, especially near saturation points, compared to locking a request into a specific backend queue early (avoiding scheduling regret). While adding another queuing layer introduces new sources of queuing latency, the goal is for these gains to offset that overhead.

### PR DESCRIPTION
**Pull request by LukeAVanDrie**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This PR introduces an experimental Flow Control layer to enhance request admission control, providing more sophisticated queuing, fairness, and multi-tenancy capabilities beyond the current saturation-based load shedding.

The key changes include:

1.  **Feature Flag**: All new flow control functionality is gated behind the `ENABLE_EXPERIMENTAL_FLOW_CONTROL_LAYER` environment variable. When disabled, the existing admission behavior is preserved.

2.  **Admission Control Refactoring**:
    *   An `AdmissionController` interface has been introduced in `pkg/epp/requestcontrol` to abstract admission control logic.
    *   `LegacyAdmissionController`: Implements the existing saturation-based shedding.
    *   `FlowControlAdmissionController`: Integrates with the new Flow Control system. This controller includes logic to adapt requests to the Flow Control interface and translate outcomes.
    *   The `Director` now uses the `AdmissionController` interface, making the admission strategy pluggable.

3.  **Component Wiring**:
    *   `cmd/epp/runner/runner.go` initializes the `FlowController` and `FlowRegistry` when the feature flag is active.
    *   The `runner` instantiates and injects the appropriate `AdmissionController` (either `Legacy` or `FlowControl`) into the `Director`.

4.  **Configuration**:
    *   The Flow Control system is currently initialized with a hardcoded default configuration in `runner.go`. Follow-up PRs will address dynamic configuration.

5.  **Testing**:
    *   Comprehensive table-driven unit tests have been added in `pkg/epp/requestcontrol/admission_test.go` to cover both `LegacyAdmissionController` and `FlowControlAdmissionController` behaviors, including edge cases and outcome translations.
    *   Tests for the `flowControlRequest` adapter have been included.
    *   `pkg/epp/requestcontrol/director_test.go` has been refactored to mock the `AdmissionController` interface, focusing tests on the `Director`'s orchestration logic.
    *   Integration tests in `test/integration/epp/hermetic_test.go` are updated to use the `LegacyAdmissionController` to ensure no change in default behavior.
    *   All test files have been polished for clarity and adherence to project standards.

This refactoring improves the design's modularity, testability, and extensibility, allowing for easier development and maintenance of admission control strategies.

**Which issue(s) this PR fixes**:
Fixes #674 

**Does this PR introduce a user-facing change?**:
```release-note
feat: Introduce experimental flow control layer gated by the `ENABLE_EXPERIMENTAL_FLOW_CONTROL_LAYER` feature flag.
```

### CODE REVIEW

File: test/integration/epp/hermetic_test.go
Comment (LukeAVanDrie) [inline test/integration/epp/hermetic_test.go:1]: Waiting on my test isolation changes in #1627 to land before I add the new FC cases. Will send these out in a separate PR. Right now, I want to unblock others from experimenting with the component. I have of course deployed and benchmarked this to verify the integration is working as expected.

File: cmd/epp/runner/runner.go [hunk 1]
```diff
@@ -68,11 +71,25 @@ import (
...
-	// used as feature flag for the pluggable data layer.
+	// enableExperimentalDatalayerV2 defines the environment variable used as feature flag for the pluggable data layer.
 	enableExperimentalDatalayerV2 = "ENABLE_EXPERIMENTAL_DATALAYER_V2"
+	// enableExperimentalFlowControlLayer defines the environment variable used as a feature flag for the pluggable flow
+	// control layer.
+	enableExperimentalFlowControlLayer = "ENABLE_EXPERIMENTAL_FLOW_CONTROL_LAYER"
 )
 
+// TODO: this is hardcoded for POC only. This needs to be hooked up to our text-based config story.
+var flowControlConfig = flowcontrol.Config{
```
Comment (LukeAVanDrie) [inline cmd/epp/runner/runner.go:82]: This is admittedly sloppy. I will send out follow PRs to align this with how the rest of our pluggable layers are configured (e.g., the text-based config). For now, I am prioritizing getting the POC checked in with a functional e2e story.

File: pkg/epp/requestcontrol/director.go [hunk 2]
```diff
@@ -59,15 +60,29 @@ type SaturationDetector interface {
...
 }
 
+// flowController defines the minimal interface required by the Director for flow control.
+type flowController interface {
+	EnqueueAndWait(req types.FlowControlRequest) (types.QueueOutcome, error)
+}
+
 // NewDirectorWithConfig creates a new Director instance with all dependencies.
-func NewDirectorWithConfig(datastore Datastore, scheduler Scheduler, saturationDetector SaturationDetector, config *Config) *Director {
+func NewDirectorWithConfig(
```
Comment (kfswain) [inline pkg/epp/requestcontrol/director.go:58]: very minor nit: can we put the 2 flow control vars next to each other?

File: pkg/epp/requestcontrol/director.go [hunk 3]
```diff
@@ -323,3 +356,53 @@ func (d *Director) runPostResponsePlugins(ctx context.Context, request *scheduli
 		loggerDebug.Info("Completed running post-response plugin successfully", "plugin", plugin.TypedName())
 	}
 }
+
+// --- Flow Control Integration ---
+
+// flowControlRequest is an adapter that implements the types.FlowControlRequest interface, wrapping the director's
+// internal request context.
+type flowControlRequest struct {
```
Comment (kfswain) [inline pkg/epp/requestcontrol/director.go]: I think we should pull all this to another file within the package.

**Comment by kfswain** [comment]
Minor stylistic comments, overall looks good! Pumped to see this integrated!

/approve

File: pkg/epp/requestcontrol/director.go (hunk 3 above)
Comment (LukeAVanDrie) [reply pkg/epp/requestcontrol/director.go]: I'm considering sending a followup where I hide this (and some of the logic on the request path) behind an `AdmissionController` type.

We could then pass `LegacyAdmissionController` or `FlowControlAdmissionController` instances into the director. Thoughts?

Comment (LukeAVanDrie) [reply pkg/epp/requestcontrol/director.go]: I've pushed this as a second commit on this PR. If you would like, I can split it off into a separate PR. Let me know whatever you prefer.

File: cmd/epp/runner/runner.go (hunk 1 above)
Comment (ahg-g) [reply cmd/epp/runner/runner.go:82]: pls create issues to track the followup items

File: pkg/epp/requestcontrol/admission.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+// rejectIfSheddableAndSaturated checks if a request should be immediately rejected because it's sheddable
+// (priority < 0) and the system is saturated.
+func rejectIfSheddableAndSaturated(
+	ctx context.Context,
+	sd saturationDetector,
+	reqCtx *handlers.RequestContext,
+	candidatePods []backendmetrics.PodMetrics,
+	priority int,
+) error {
+	if priority < 0 {
```
Comment (ahg-g) [inline pkg/epp/requestcontrol/admission.go]: instead of baking this assumption here, pls define a "IsSheddable" request function, perhaps under https://github.com/kubernetes-sigs/gateway-api-inference-extension/tree/main/pkg/epp/util/request

File: pkg/epp/requestcontrol/admission.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+	if err := rejectIfSheddableAndSaturated(ctx, fcac.saturationDetector, reqCtx, candidatePods, priority); err != nil {
+		return err
+	}
+
+	logger.V(logutil.TRACE).Info("Request proceeding to flow control", "requestID", reqCtx.SchedulingRequest.RequestId)
+	fairnessID := reqCtx.FairnessID
+	if fairnessID == "" {
+		logger.V(logutil.TRACE).Info("No fairnessID provided, using default",
+			"requestID", reqCtx.SchedulingRequest.RequestId, "defaultFairnessID", defaultFairnessID)
+		fairnessID = defaultFairnessID
```
Comment (ahg-g) [inline pkg/epp/requestcontrol/admission.go]: I think defaulting should be applied earlier when the reqCtx is created, not downstream here.

File: pkg/epp/requestcontrol/admission.go (hunk 4 above)
Comment (LukeAVanDrie) [reply pkg/epp/requestcontrol/admission.go]: Done, see the latest commit on this PR.

File: pkg/epp/requestcontrol/admission.go (hunk 5 above)
Comment (LukeAVanDrie) [reply pkg/epp/requestcontrol/admission.go]: Good call, I've updated this in the latest commit in the PR.
